shared/src/main/scala/mlscript/ConstraintSolver.scala

```diff
@@ -279,26 +286,108 @@ class ConstraintSolver extends NormalForms { self: Typer =>
         }
       }
     
-    def rec(lhs: SimpleType, rhs: SimpleType, sameLevel: Bool)
+    /**
+      * Recursively constrain the given simple types. While constraining
+      * keep tracking the provenance of the types in `cctx`. In case of
+      * a constraining error chain can be used to generate error messages
+      * and show the flow of types that lead to the error.
+      * 
+      * @param lhs
+      * @param rhs
+      * @param nestedProv
+      * @param raise
+      * @param cctx
+      */
+    def rec(lhs: SimpleType, rhs: SimpleType, nestedProv: Opt[NestedTypeProvenance] = N)
           (implicit raise: Raise, cctx: ConCtx): Unit = {
       constrainCalls += 1
-      // Thread.sleep(10)  // useful for debugging constraint-solving explosions debugged on stdout
-      recImpl(lhs, rhs)(raise,
-        if (sameLevel)
-          (if (cctx._1.headOption.exists(_ is lhs)) cctx._1 else lhs :: cctx._1)
+      var sameLhs = cctx._1.headOption.exists(_.prov is lhs.prov)
+      var sameRhs = cctx._2.headOption.exists(_.prov is rhs.prov)
+      
+      // check if these tuples are implicit
+      // don't add them to chain
+      (lhs, rhs) match {
+        case (lt: TupleType, rt: TupleType) =>
+          if (lt.implicitTuple || rt.implicitTuple)
+            sameLhs = true
+            sameRhs = true
+        case _ => ()
+      }
```

Comment:
Instead of this hack, wouldn't it be better to simply not associate any provenance to tuples types derived from implicit tuples?

Reply:
Yes provenance is no longer associated with implicit tuples. I missed removing this old check.

Reply:
Ok, cool.